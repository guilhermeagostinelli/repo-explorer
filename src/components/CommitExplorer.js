import React, { useState, useEffect, Fragment } from "react";
import UnorderedList from "./UnorderedList";
import ListItem from "./ListItem";
import SearchField from "./SearchField";
import Commit from "./Commit";
import { getCommits } from "../api/gitHub";

const CommitExplorer = ({ match }) => {
  const user = match.params.user;
  const repo = match.params.repo;
  const [commitQuery, setCommitQuery] = useState("");
  const [commits, setCommits] = useState([]);

  useEffect(() => {
    // Get list of commits by user and repo.
    (async () => {
      try {
        setCommits(await getCommits(user, repo, 1));
      } catch (err) {
        if (err.message !== "Not Found") {
          console.error(err);
          alert("Error fetching results from GitHub's API...");
        }
        setCommits([]);
      }
    })();
  }, [user, repo]);

  // Filters by commit message.
  const filteredCommits = commits.filter(c =>
    c.commit.message.toLowerCase().includes(commitQuery.toLowerCase())
  );

  return (
    <Fragment>
      <h1>Commit Explorer</h1>
      <p>
        <SearchField
          label="Search by commit message:"
          value={commitQuery}
          onChange={setCommitQuery}
        />
      </p>
      <p>Showing {filteredCommits.length} commit(s)</p>
      <UnorderedList>
        {filteredCommits.map(c => {
          const props = {
            author: c.commit.author.name,
            comitted_at: c.commit.author.date,
            message: c.commit.message,
            html_url: c.html_url
          };
          return (
            <ListItem key={c.sha}>
              <Commit {...props}></Commit>
            </ListItem>
          );
        })}
      </UnorderedList>
    </Fragment>
  );
};

export default CommitExplorer;
