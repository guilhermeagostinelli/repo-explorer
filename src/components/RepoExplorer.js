import React, { useState, useEffect, Fragment } from "react";
import UnorderedList from "./UnorderedList";
import ListItem from "./ListItem";
import SearchField from "./SearchField";
import SortField from "./SortField";
import Repo from "./Repo";
import { getRepos } from "../api/gitHub";

const RepoExplorer = () => {
  const [userQuery, setUserQuery] = useState("reactjs");
  const [repos, setRepos] = useState([]);
  const [sortBy, setSortBy] = useState("name");
  const [sortType, setSortType] = useState("asc");

  useEffect(() => {
    // Get list of repositories by user.
    (async () => {
      try {
        setRepos(await getRepos(userQuery));
      } catch (err) {
        if (err.message !== "Not Found") {
          console.error(err);
          alert("Error fetching results from GitHub's API...");
        }
        setRepos([]);
      }
    })();
  }, [userQuery]);

  const orderedRepos = repos.sort((a, b) => {
    let t;
    if (typeof a[sortBy] === "string" && typeof b[sortBy] === "string")
      t = a[sortBy].localeCompare(b[sortBy]);
    else t = a[sortBy] > b[sortBy] ? 1 : -1;

    return t * (sortType === "asc" ? 1 : -1);
  });

  return (
    <Fragment>
      <h1>Repository Explorer</h1>
      <p>
        <SearchField
          label="Search by GitHub user:"
          value={userQuery}
          onChange={setUserQuery}
        />
      </p>
      <p>
        <SortField
          valueNamePairs={[
            { value: "name", name: "Name" },
            { value: "description", name: "Description" },
            { value: "created_at", name: "Creation date" },
            { value: "language", name: "Language" },
            { value: "stargazers_count", name: "Stars" }
          ]}
          sortByValue={sortBy}
          sortTypeValue={sortType}
          onChangeSortByCb={setSortBy}
          onChangeSortTypeCb={setSortType}
        />
      </p>
      <p>Showing {orderedRepos.length} repo(s)</p>
      <UnorderedList>
        {orderedRepos.map(r => (
          <ListItem key={r.id}>
            <Repo user={userQuery} {...r}></Repo>
          </ListItem>
        ))}
      </UnorderedList>
    </Fragment>
  );
};

export default RepoExplorer;
