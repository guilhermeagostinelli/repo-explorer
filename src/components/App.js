import React from "react";
import { BrowserRouter as Router, Switch, Route } from "react-router-dom";
import styled from "styled-components";
import RepoExplorer from "./RepoExplorer";
import CommitExplorer from "./CommitExplorer";
import NotFound from "./NotFound";

const Div = styled.div`
  text-align: center;
`;

const App = () => {
  return (
    <Div>
      <Router>
        <Switch>
          <Route exact path="/" component={RepoExplorer} />
          <Route path="/commits/:user/:repo" component={CommitExplorer} />
          <Route component={NotFound} />
        </Switch>
      </Router>
    </Div>
  );
};

export default App;
