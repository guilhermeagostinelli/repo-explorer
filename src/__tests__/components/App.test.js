import React from "react";
import { create } from "react-test-renderer";
import { shallow } from "enzyme";
import { Route } from "react-router-dom";
import "jest-styled-components";
import App from "../../components/App";
import RepoExplorer from "../../components/RepoExplorer";
import CommitExplorer from "../../components/CommitExplorer";
import NotFound from "../../components/NotFound";

jest.mock("../../components/RepoExplorer", () => "div");

describe("<App />", () => {
  it("should match the snapshot", () => {
    const app = create(<App />);
    expect(app.toJSON()).toMatchSnapshot();
  });

  it("renders routes correctly", () => {
    const wrapper = shallow(<App />);
    const pathMap = wrapper.find(Route).reduce((pathMap, route) => {
      const routeProps = route.props();
      pathMap[routeProps.path] = routeProps.component;
      return pathMap;
    }, {});
    expect(pathMap["/"]).toBe(RepoExplorer);
    expect(pathMap["/commits/:user/:repo"]).toBe(CommitExplorer);
    expect(pathMap["undefined"]).toBe(NotFound);
  });
});
