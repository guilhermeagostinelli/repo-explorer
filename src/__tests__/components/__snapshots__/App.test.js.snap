// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should match the snapshot 1`] = `
.c0 {
  text-align: center;
}

<div
  className="c0"
>
  <div
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
    location={
      Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      }
    }
    match={
      Object {
        "isExact": true,
        "params": Object {},
        "path": "/",
        "url": "/",
      }
    }
  />
</div>
`;
